### Section 1: Python Operators and Control Structures (5 questions)

#### Question 1: Write a Python program to calculate the sum of all numbers in a list using a for loop.


In [19]:
# List of numbers
numbers = [1, 2, 3, 4, 5]
# variable to store the sum
total = 0
# Loop to iterate and add each number to total
for num in numbers:
    total += num
    
# print final result
print(total)

15


Explanation:
* 1.We initialize total = 0 to store the sum.
* 2.We loop through the numbers list using a for loop.
* 3.In each iteration, we add the current number to total.
* 4.After the loop completes, we print the final sum.

#### Question 2: What is the output of the following code?

In [20]:
x = 5
y = 2
print(x ** y)

25


Why?
* 1.** is the exponentiation operator in Python.
* 2.x ** y means 5 raised to the power of 2 → 5^2 = 25.

#### Question 3: Write a Python program to check if a number is a palindrome using a while loop.

A palindrome is a number that remains the same when reversed.
For example:

12321 → Palindrome ✅
12345 → Not a Palindrome ❌

In [21]:
def is_palindrome(num):
    original = num
    reversed_num = 0
    temp = num
    
    while temp > 0:
        digit = temp % 10
        reversed_num = reversed_num * 10 + digit
        temp //= 10
        
    return original == reversed_num 
num = 12321
print(is_palindrome(num))

True


Explanation:
* 1.Extract digits from the number one by one using the modulus (%) operator.
* 2.Reverse the number by multiplying the existing reversed number by 10 and adding the extracted digit.
* 3.Compare the original number with the reversed number.
* 4.If they are equal, return True (Palindrome); otherwise, return False.

#### Question 4: What is the output of the following code?

<pre>fruits = ['apple', 'banana', 'cherry']
print(fruits[-1])</pre>


In [22]:
fruit = ['apple', 'banana', 'cherry']
print(fruit[-1])

cherry


Why?
* 1.In Python, negative indexing allows you to access elements from the end of a list.
* 2.fruits[-1] refers to the last element of the list.
* 3.The last element in ['apple', 'banana', 'cherry'] is 'cherry'.

#### Question 5: Write a Python program to calculate the average of all numbers in a list using a for loop.

In [23]:
numbers = [1, 2, 3, 4, 5]
total = 0
for num in numbers:
    total += num
    
average = total / len(numbers)
print(average)

3.0


Explanation:
* 1.Initialize total = 0 to store the sum.
* 2.Loop through the numbers list and add each number to total.
* 3.Divide the sum by the length of the list using len(numbers).
* 4.Print the result to get the average.

### Section 2: Lists and Dictionaries (5 questions)

#### Question 6: Write a Python program to create a dictionary with student names as keys and their grades as values.

In [24]:
students = ['John', 'Alice', 'Bob']
grades = [90, 85, 95]
student_dict = dict(zip(students, grades))
print(student_dict)

{'John': 90, 'Alice': 85, 'Bob': 95}


Explanation:
* zip(students, grades) pairs each student with their corresponding grade.
* dict(zip(...)) converts the paired values into a dictionary.
* The final dictionary contains student names as keys and grades as values.

#### Question 7: What is the output of the following code?
<pre>numbers = [1, 2, 3, 4, 5]
print(numbers[1:3])</pre>


In [25]:
numbers = [1, 2, 3, 4, 5]
print(numbers[1:3])

[2, 3]


Why?
* In Python, list slicing is used to extract a portion of a list.
* numbers[1:3] means:
* Start from index 1 (which is 2).
* Go up to, but not including, index 3 (which is 4).
* The extracted portion is [2, 3].

#### Question 8: Write a Python program to sort a list of numbers in ascending order using the sorted() function.

In [26]:
numbers = [4, 2, 9, 6, 5]
sorted_numbers = sorted(numbers)
print(sorted_numbers)

[2, 4, 5, 6, 9]


Explanation:
* The sorted() function returns a new sorted list without modifying the original list.
* Since the numbers are initially unordered, sorted(numbers) sorts them in ascending order.
* The sorted list is stored in sorted_numbers and then printed.

#### Question 9: What is the output of the following code?
<pre>fruits = {'apple': 5, 'banana': 10, 'cherry': 15} 
print(fruits.get('banana'))</pre>


In [27]:
fruits = {'apple' : 5, 'banana': 10, 'cherry' : 15}
print(fruits.get('banana'))

10


Why?
* Dictionaries in Python store key-value pairs.
* The .get('banana') method retrieves the value associated with the key 'banana'.
* The key 'banana' exists in the dictionary with a value of 10.

#### Question 10: Write a Python program to create a list of squares of all numbers from 1 to 10 using a list comprehension.

In [28]:
squares = [x ** 2 for x in range(1, 11)]
print(squares)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


Explanation:
* List comprehension [x**2 for x in range(1, 11)] generates a list of squares.
* range(1, 11) iterates over numbers from 1 to 10.
* Each number x is squared (x**2) and added to the lis

### Section 3: Advanced Topics (5 questions)

#### Question 11: Write a Python program to calculate the factorial of a number using recursion.

In [29]:
def factorial(n):
    if n == 0 or n == 1:
        return 1
    else:
        return n * factorial(n - 1)
    
num = 5
print(factorial(num))

120


Explanation:
* The base case:
If n == 0 or n == 1, return 1 (since 0! = 1! = 1).<br>
The recursive case:
* Multiply n by factorial(n - 1), which keeps calling itself until reaching the base case.<br>
* Example for n = 5:<br>
factorial(5) = 5 * factorial(4)<br>
factorial(4) = 4 * factorial(3)<br>
factorial(3) = 3 * factorial(2)<br>
factorial(2) = 2 * factorial(1)<br>
factorial(1) = 1  (Base case)<br>

* 5 * 4 * 3 * 2 * 1 = 120

#### Question 12: What is the output of the following code?
<pre>numbers = [1, 2, 3, 4, 5]
print(list(reversed(numbers)))
</pre>

In [30]:
numbers = [1, 2, 3, 4, 5]
print(list(reversed(numbers)))

[5, 4, 3, 2, 1]


Why?
* The reversed() function returns an iterator that iterates over the list in reverse order.
* list(reversed(numbers)) converts that iterator into a list.
* The original list [1, 2, 3, 4, 5] is reversed to [5, 4, 3, 2, 1]

#### Question 13: Write a Python program to create a generator that yields the Fibonacci sequence up to the nth term.


In [32]:
def fibonacci_generator(n):
    a, b = 0, 1
    for _ in range(n):
        yield a
        a, b = b, a + b
        
n = 10
print(list(fibonacci_generator(n)))

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]


Explanation of Code Execution:

* yield a: Returns the current Fibonacci number but pauses execution.
* a, b = b, a + b: Moves to the next Fibonacci number.
* The loop runs n times, yielding n Fibonacci numbers.

#### Question 14: What is the output of the following code?
<pre>import math
print(math.pi)
</pre>

In [33]:
import math
print(math.pi)

3.141592653589793


Why?
* The math module in Python provides mathematical functions and constants.
* math.pi returns the mathematical constant π (pi), which is approximately 3.141592653589793.
* This value is commonly used in geometry and trigonometry

#### Question 15: Write a Python program to print an inverted star pyramid of height 5.
<pre>*********
 *******
  *****
   ***
    *
</pre>

In [34]:
def inverted_pyramid(height):
    for i in range(height, 0, -1):
        spaces = ' ' * (height - i)
        stars = '*' * (2 * i -1)
        print(spaces + stars)
        
height = 5
inverted_pyramid(height)

*********
 *******
  *****
   ***
    *


Your Markdown structure is mostly correct, but I made some improvements for better readability and proper formatting. Here's the corrected version:  

```markdown
### Explanation:
1. **Loop from height down to 1** (`range(height, 0, -1)`)
   - Starts at `height = 5`, decreases to `1`.

2. **Calculate spaces** (`' ' * (height - i)`)
   - Number of spaces increases as `i` decreases.
   - Example:
     ```
     Row 1: 0 spaces
     Row 2: 1 space
     Row 3: 2 spaces
     Row 4: 3 spaces
     Row 5: 4 spaces
     ```

3. **Calculate stars** (`'*' * (2 * i - 1)`)
   - Number of stars follows the pattern:
     ```
     Row 1: 9 stars
     Row 2: 7 stars
     Row 3: 5 stars
     Row 4: 3 stars
     Row 5: 1 star
     ```

4. **Print `spaces + stars` to form the pyramid shape.**

### **Final Output:**
```
<pre>*********
 *******
  *****
   ***
    *</pre>
```
```